### PR TITLE
Prevent crashes when an iuse doesn't exist.

### PR DESCRIPTION
**Pull request by kevingranade**
#### Summary
SUMMARY: None

#### Purpose of change
The recent removal of the ROYAL_JELLY iuse in #39782 triggered breakage in downstream mods that invoked it.
See https://discourse.cataclysmdda.org/t/segmentation-fault-on-finalizing-items/23367
The root cause is that the code handling iuse function lookup and assignment to itypes was generating a null "use_function" object, but was then not checking it for null before assigning it to the iuse, and it would inevitably be invoked during finalization, causing a crash.

#### Describe the solution
Adds a wrapper around iuse function lookup and assignment that refrains from emplacing the use_function if it is null.  The lookup function already reports the failed lookup, so no further diagnostics are needed.

#### Describe alternatives you've considered
I considered trying to make the returned null object capable of being dereferenced, but that seems like a bandaid type fix that might be fragile in the future, and refraining from emplacing the function object is more robust.

#### Testing
Create an item that specifies a non-existent iuse id and load it.
It should display a debug message but continue to load.

### CODE REVIEW

File: src/item_factory.cpp [hunk 1]
```diff
@@ -2845,6 +2845,15 @@ void Item_factory::set_use_methods_from_json( const JsonObject &jo, const std::s
     }
 }
 
+// Helper to safely look up and store iuse actions.
+void Item_factory::emplace_usage( std::map<std::string, use_function> &container, std::string iuse_id )
```
Comment (ZhilkinSerg) [inline src/item_factory.cpp:2849]:
```suggestion
void Item_factory::emplace_usage( std::map<std::string, use_function> &container,
                                  const std::string &iuse_id )
```

File: src/item_factory.h [hunk 2]
```diff
@@ -301,6 +301,8 @@ class Item_factory
         void load( relic &slot, const JsonObject &jo, const std::string &src );
 
         //json data handlers
+        void emplace_usage( std::map<std::string, use_function> &container, std::string iuse_id );
```
Comment (ZhilkinSerg) [inline src/item_factory.h:304]:
```suggestion
        void emplace_usage( std::map<std::string, use_function> &container, const std::string &iuse_id );
```

**Comment by ZhilkinSerg** [comment]
Legacy use actions (e.g. `"use_action": "RPGDIE"`) with invalid data are reported and loading of game continues after that:

```
 DEBUG    : Received unrecognized iuse function RPGDIEZZZs, using iuse::none instead

 FUNCTION : use_function Item_factory::usage_from_string(const string&) const
 FILE     : src/item_factory.cpp
 LINE     : 2888
```

But modern use actions (e.g. `"use_action": { "type": "bandolier", "capacity": 20, "ammo": [ "arrow", "bolt" ], "draw_cost": 20 }`) with invalid data are reported and loading of game fails:

```
 DEBUG    : Received unrecognized iuse function bandolierXXX, using iuse::none instead

 FUNCTION : use_function Item_factory::usage_from_string(const string&) const
 FILE     : src/item_factory.cpp
 LINE     : 2888
```

```
 DEBUG    : Error: data/mods//dda/../../json/items/armor/ammo_pouch.json: line 259:28: unknown use_action

    "encumbrance": 3,
    "material_thickness": 1,
    "use_action": { "type":
                           ^
                            "bandolierXXX", "capacity": 20, "ammo": [ "arrow", "bolt" ], "draw_cost": 20 },
    "flags": [ "WAIST", "OVERSIZE", "WATER_FRIENDLY" ]
  },


 FUNCTION : bool main_menu::load_character_tab(bool)
 FILE     : src/main_menu.cpp
 LINE     : 1103

```

**Comment by kevingranade** [comment]
I'll look at those too, I thought they were handled better.

**Comment by kevingranade** [comment]
So there's a tradeoff, if usage_from_object() doesn't throw an exception in this situation, we don't get as good of an error message when it fails, but I prefer the robustness for this situation.

**Comment by jbytheway** [comment]
> So there's a tradeoff, if usage_from_object() doesn't throw an exception in this situation, we don't get as good of an error message when it fails, but I prefer the robustness for this situation.

Is it worth throwing an error in `test_mode`?

**Comment by kevingranade** [comment]
> Is it worth throwing an error in test_mode?

Fantastic idea.

**Pull request merged**